modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/db/wal/reader/IgniteWalReaderTest.java

```diff
@@ -996,6 +1056,189 @@ private void runRemoveOperationTest(CacheAtomicityMode mode) throws Exception {
             deletesFound != null && deletesFound > 0);
     }
 
+    /**
+     * Tests transaction generation and WAL for putAll cache operation.
+     *
+     * @throws Exception if failed.
+     */
+    @Test
+    public void testPrimaryFlagOnTwoNodes() throws Exception {
```

Comment:
we still need test streaming on unstable topology with the assertion that streamed entries cnt == cnt recorded as primary.
single and multi-key tx should be checked as well as atomics.